backends/dpdk/dpdkArch.cpp

```diff
@@ -3035,4 +3035,153 @@ std::vector<std::pair<cstring, const IR::Type *>>
 cstring DpdkAddPseudoHeaderDecl::pseudoHeaderInstanceName;
 cstring DpdkAddPseudoHeaderDecl::pseudoHeaderTypeName;
 
+// create and add register declaration instance to program
+IR::IndexedVector<IR::StatOrDecl> *InsertReqDeclForIPSec::addRegDeclInstance(
+    std::vector<cstring> portRegs) {
+    auto decls = new IR::IndexedVector<IR::StatOrDecl>;
+    for (auto reg : portRegs)
+        decls->push_back(getRegDeclarationInstance(reg, IPSEC_PORT_REG_SIZE,
+                                                   IPSEC_PORT_REG_INDEX_BITWIDTH,
+                                                   IPSEC_PORT_REG_INITVAL_BITWIDTH));
+    return decls;
+}
+
+const IR::Node *InsertReqDeclForIPSec::preorder(IR::P4Program *program) {
+    auto ipsecInfo = new CollectIPSecInfo(is_ipsec_used, sa_id_width, refMap, typeMap, structure);
+    ipsecInfo->setCalledBy(this);
+    program->apply(*ipsecInfo);
+    if (!is_ipsec_used) return program;
+    cstring resName = "";
+    if (!uniqueNames(refMap, registerInstanceNames, resName))
+        ::error(ErrorType::ERR_RESERVED, "%1% name is reserved for DPDK IPSec port register",
+                resName);
+    bool ipsecHdrAdded = false;
+    bool ipsecPortRegAdded = false;
+    auto new_objs = new IR::Vector<IR::Node>;
+
+    // Create platform_hdr structure with just one field to hold the security association index
+    IR::ID newHeaderFieldName("sa_id");
+    IR::IndexedVector<IR::StructField> newHeaderFields;
+    newHeaderFields.push_back(
+        new IR::StructField(newHeaderFieldName, IR::Type_Bits::get(sa_id_width)));
+    if (!uniqueNames(refMap, {newHeaderName}, resName)) {
+        ::error(ErrorType::ERR_RESERVED, "%1% type name is reserved for DPDK platform header",
+                newHeaderName);
+        return program;
+    }
+    ipsecHeader = new IR::Type_Header(IR::ID(newHeaderName), newHeaderFields);
+
+    // Programs using ipsec accelerator must contain ethernet and ipv4 headers, hence a struct
+    // containing headerinstances for all headers must be present
+    BUG_CHECK(structure->header_type != "",
+              "Encapsulating struct containing all headers used in the program is missing");
+    for (auto obj : program->objects) {
```

Comment:
why not use the visitor methods for this?

Reply:
Visitors for Type_Struct and P4Parser? This is for adding the new  header and register declarations to the program. Only for placing these new declarations better in the program, I have added check for Type_Struct and Parser.